File: src/main/kotlin/me/hippodev/protocol/StatusJson.kt
package me.hippodev.protocol

import me.hippodev.config.FallbackStatus

fun buildFallbackJson(fallback: FallbackStatus): String {
    val sb = StringBuilder()
    sb.append('{')
    sb.append("\"version\":{")
    sb.append("\"name\":\"").append(escapeJson(fallback.version.name)).append("\",")
    sb.append("\"protocol\":").append(fallback.version.protocol)
    sb.append("},")
    // Previously this dropped fallback.motd in as flat, unformatted text - no legacy &-codes, no
    // MiniMessage. Now it goes through the same rich-text pipeline as every other message (see
    // TextFormat.kt), so &-codes and MiniMessage tags both work in the MOTD too.
    sb.append("\"description\":").append(toJsonComponent(fallback.motd))
    fallback.players?.let {
        sb.append(",\"players\":{\"online\":").append(it.online)
            .append(",\"max\":").append(it.max).append('}')
    }
    fallback.faviconDataUri?.let {
        sb.append(",\"favicon\":\"").append(escapeJson(it)).append('"')
    }
    sb.append('}')
    return sb.toString()
}

private fun escapeJson(value: String): String {
    val sb = StringBuilder(value.length)
    for (c in value) {
        when (c) {
            '"' -> sb.append("\\\"")
            '\\' -> sb.append("\\\\")
            '\n' -> sb.append("\\n")
            '\r' -> sb.append("\\r")
            '\t' -> sb.append("\\t")
            else -> if (c.code < 0x20) sb.append("\\u%04x".format(c.code)) else sb.append(c)
        }
    }
    return sb.toString()
}

File: src/status_json.rs
//! Port of `protocol/StatusJson.kt`.

use crate::config::FallbackStatus;
use crate::text_format::{escape_json, to_json_component};
use std::fmt::Write as _;

pub fn build_fallback_json(fallback: &FallbackStatus) -> String {
    let mut out = String::new();
    out.push('{');
    out.push_str("\"version\":{");
    let _ = write!(out, "\"name\":\"{}\",", escape_json(&fallback.version.name));
    let _ = write!(out, "\"protocol\":{}", fallback.version.protocol);
    out.push_str("},");
    // Goes through the same rich-text pipeline as every other message (see text_format.rs), so
    // &-codes and MiniMessage tags both work in the MOTD too.
    out.push_str("\"description\":");
    out.push_str(&to_json_component(&fallback.motd));
    if let Some(players) = &fallback.players {
        let _ = write!(out, ",\"players\":{{\"online\":{},\"max\":{}}}", players.online, players.max);
    }
    if let Some(favicon) = &fallback.favicon_data_uri {
        let _ = write!(out, ",\"favicon\":\"{}\"", escape_json(favicon));
    }
    out.push('}');
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::{PlayersInfo, VersionInfo};

    #[test]
    fn builds_expected_shape() {
        let fallback = FallbackStatus {
            motd: "&eHello".to_string(),
            version: VersionInfo { name: "1.21".to_string(), protocol: 767 },
            players: Some(PlayersInfo { online: 3, max: 20 }),
            favicon_data_uri: Some("data:image/png;base64,abc".to_string()),
        };
        let json = build_fallback_json(&fallback);
        assert!(json.contains(r#""name":"1.21""#));
        assert!(json.contains(r#""protocol":767"#));
        assert!(json.contains(r#""online":3"#));
        assert!(json.contains(r#""max":20"#));
        assert!(json.contains(r#""favicon":"data:image/png;base64,abc""#));
        assert!(json.contains(r#""color":"yellow""#)); // &e rendered through the rich-text pipeline
    }

    #[test]
    fn omits_optional_fields_when_absent() {
        let fallback = FallbackStatus {
            motd: "hi".to_string(),
            version: VersionInfo { name: "1.21".to_string(), protocol: 767 },
            players: None,
            favicon_data_uri: None,
        };
        let json = build_fallback_json(&fallback);
        assert!(!json.contains("players"));
        assert!(!json.contains("favicon"));
    }
}
